File: src/Functional/OCaml/patterns/mediator.ml
exception Unknown_colleague of string

type receiver = sender:string -> message:string -> string

type mediator = (string, receiver) Hashtbl.t

let register mediator name receiver = Hashtbl.replace mediator name receiver

let send mediator ~sender ~recipient ~message =
  match Hashtbl.find_opt mediator recipient with
  | Some receiver -> receiver ~sender ~message
  | None -> raise (Unknown_colleague recipient)

let () =
  let mediator = Hashtbl.create 2 in
  register mediator "payment" (fun ~sender ~message ->
      Printf.sprintf "payment received %s from %s" message sender);
  register mediator "inventory" (fun ~sender ~message ->
      Printf.sprintf "inventory received %s from %s" message sender);

  assert (
    send mediator ~sender:"payment" ~recipient:"inventory" ~message:"paid"
    = "inventory received paid from payment");
  assert (
    send mediator ~sender:"inventory" ~recipient:"payment" ~message:"reserved"
    = "payment received reserved from inventory");

  let rejected_unknown =
    try
      ignore (send mediator ~sender:"payment" ~recipient:"shipping" ~message:"paid");
      false
    with Unknown_colleague "shipping" -> true
  in
  assert rejected_unknown

File: src/Functional/Haskell/patterns/mediator.hs
module Main where

import qualified Data.Map.Strict as Map

type Colleague = String -> String -> String

data CheckoutMediator = CheckoutMediator (Map.Map String Colleague)

register :: String -> Colleague -> CheckoutMediator -> CheckoutMediator
register name receiver (CheckoutMediator colleagues) =
  CheckoutMediator (Map.insert name receiver colleagues)

send :: CheckoutMediator -> String -> String -> String -> Either String String
send (CheckoutMediator colleagues) sender recipient message =
  case Map.lookup recipient colleagues of
    Just receiver -> Right (receiver sender message)
    Nothing -> Left ("unknown colleague: " ++ recipient)

payment :: Colleague
payment sender message = "payment<-" ++ sender ++ ":" ++ message

inventory :: Colleague
inventory sender message = "inventory<-" ++ sender ++ ":" ++ message

verifyMediator :: Bool
verifyMediator =
  let mediator =
        register "inventory" inventory $
          register "payment" payment $
            CheckoutMediator Map.empty
   in send mediator "payment" "inventory" "paid"
        == Right "inventory<-payment:paid"
        && send mediator "inventory" "payment" "reserved"
          == Right "payment<-inventory:reserved"
        && send mediator "payment" "shipping" "dispatch"
          == Left "unknown colleague: shipping"

main :: IO ()
main =
  if verifyMediator
    then putStrLn "Haskell Mediator: passed"
    else error "Haskell Mediator failed"
